src/sail_lean_backend/Sail/IntRange.lean

```diff
@@ -0,0 +1,67 @@
+structure IntRange where
+  start : Int := 0
+  stop  : Int
+  step  : Nat := 1
+  step_pos : 0 < step := by omega
+
+namespace IntRange
+
+instance instMemIntRange : Membership Int IntRange where
+  mem r i := r.start ≤ i ∧ i < r.stop ∧ (i - r.start).toNat % r.step = 0
+
+instance : DecidableRel (fun (i : Int) (r : IntRange) => i ∈ r) := fun i r => by
+  simp [instMemIntRange]
+  infer_instance
+
+/-- The number of elements in the range. -/
+def size (r : IntRange) : Nat := ((r.stop - r.start).toNat + r.step - 1) / r.step
+
+@[inline] protected def forIn' [Monad m] (range : IntRange) (init : β)
+    (f : (i : Int) → i ∈ range → β → m (ForInStep β)) : m β :=
+  if size_lt : range.start + range.step * range.size < range.stop then
```

Comment:
Why do we need this condition? This seems to be the source of the tests breaking:
```lean
def range := [0:7 + 1:1]i
#eval range.start + range.step * range.size < range.stop
```
returns false

Reply:
Yea, I think there's an off-by-one error there.

We need it on the loop to prove the `i ∈ range` in the recursive call. The `numIterations` in the loop is needed to avoid Lean switching to wellfounded recursion.

Reply:
I had a quick glance and the `Nat` version doesn't need this guard. What's the reason?

Reply:
I could also first do it without structural recursion and prove the termination manually

Reply:
What's wrong with wellfounded-recursion?

Reply:
It's currently near impossible to show anything about functions defined by wellfounded recursion. (Though afaik that's bound to change)

Reply:
Really? In my automata dev I used the function induction scheme and it was quite alright I thought

Reply:
Maybe it's better now, not sure when I last tried, but it might be before the function induction schemes were added.